### PR TITLE
[MAINTENANCE] Improve `DatasourceDict` performance

### PR DESCRIPTION
**Pull request by cdkini**
Utilize `Store.get_all` to retrieve and build the entire collection of datasources instead of iterating and making individual GET requests

- [x] Description of PR changes above includes a link to [an existing GitHub issue](https://github.com/great-expectations/great_expectations/issues)
- [x] PR title is prefixed with one of: [BUGFIX], [FEATURE], [DOCS], [MAINTENANCE], [CONTRIB]
- [x] Code is linted - run `invoke lint` (uses `black` + `ruff`)
- [x] Appropriate tests and docs have been updated

For more information about contributing, see [Contribute](https://docs.greatexpectations.io/docs/contributing/contributing_checklist).

After you submit your PR, keep the page open and **monitor the statuses of the various checks made by our continuous integration process at the bottom of the page. Please fix any issues that come up** and [reach out on Slack](https://greatexpectations.io/slack) if you need help. Thanks for contributing!

### CODE REVIEW

File: great_expectations/data_context/store/checkpoint_store.py [hunk 1]
```diff
@@ -40,13 +40,16 @@ class CheckpointStore(ConfigurationStore):
     _configuration_class = CheckpointConfig
 
     @override
-    def ge_cloud_response_json_to_object_dict(self, response_json: Dict) -> Dict:
+    def ge_cloud_response_json_to_object_dict(
+        self, response_json: Dict
+    ) -> Union[Dict, List[Dict]]:
```
Comment (cdkini) [inline great_expectations/data_context/store/checkpoint_store.py]: I don't love this union - any way to get around this? If we want to support retrieving and parsing entire collections, we need to support a collection return type.

File: great_expectations/data_context/store/datasource_store.py [hunk 2]
```diff
@@ -126,24 +126,27 @@ def deserialize(
...
-            if len(data) > 1:
-                # TODO: handle larger arrays of Datasources
-                raise TypeError(
-                    f"GX Cloud returned {len(data)} Datasources but expected 1"
-                )
-            data = data[0]
+            if len(data) == 1:
+                data = data[
+                    0
+                ]  # If a single datasource is returned in list form, extract and treat as individual
```
Comment (cdkini) [inline great_expectations/data_context/store/datasource_store.py]: This is just keeping legacy behavior but I feel a little iffy dealing with a list when we have just one element - the backend should provide a consistent structure in my opinion (maybe always a list even if one element?)

File: great_expectations/data_context/store/in_memory_store_backend.py [hunk 3]
```diff
@@ -59,7 +59,9 @@ def _get(self, key):
 
     @override
     def _get_all(self) -> list[Any]:
-        raise NotImplementedError
+        return [
+            val for key, val in self._store.items() if key != self.STORE_BACKEND_ID_KEY
+        ]
```
Comment (cdkini) [inline great_expectations/data_context/store/in_memory_store_backend.py:64]: Implemented to get in-memory based tests working - we should get this working for all backends but not a priority

File: great_expectations/data_context/store/store.py [hunk 4]
```diff
@@ -194,6 +206,8 @@ def get(
             value = self._store_backend.get(self.key_to_tuple(key))
 
         if value:
+            if isinstance(value, list):
+                return [self.deserialize(v) for v in value]
```
Comment (cdkini) [inline great_expectations/data_context/store/store.py]: Bleh same note as above - don't love unions but this is necessary if we want to retrieve an entire collection (like we do when calling `context.datasources`)

File: great_expectations/core/datasource_dict.py [hunk 5]
```diff
@@ -78,40 +72,49 @@ def data(self) -> dict[str, FluentDatasource | BaseDatasource]:  # type: ignore[
...
+                if isinstance(config, FluentDatasource):
+                    name = config.name
+                    datasources[name] = self._init_fluent_datasource(
+                        name=name, ds=config
+                    )
+                else:
+                    name = config["name"]
+                    datasources[name] = self._init_block_style_datasource(
+                        name=name, config=config
+                    )
```
Comment (Kilo59) [inline great_expectations/core/datasource_dict.py:89]: Can we put a short lived TTL cache on this before merging this?

File: great_expectations/data_context/store/checkpoint_store.py (hunk 1 above)
Comment (Kilo59) [reply great_expectations/data_context/store/checkpoint_store.py]: Can we define a separate method for operations against whole collections?

File: great_expectations/data_context/store/gx_cloud_store_backend.py [hunk 6]
```diff
@@ -257,13 +257,6 @@ def _send_get_request_to_api(self, url: str, params: dict | None = None) -> dict
             )
             response.raise_for_status()
             response_json: dict = response.json()
-            if (
-                isinstance(response_json["data"], list)
-                and len(response_json["data"]) == 0
-            ):
-                raise StoreBackendError(
-                    "Unable to get object in GX Cloud Store Backend: Object does not exist."
-                )
```
Comment (cdkini) [inline great_expectations/data_context/store/gx_cloud_store_backend.py]: Extract this out into a separate PR

**Review by Kilo59 — APPROVED** [review]
